File: models/request.py
from datetime import datetime
from typing import Any, Dict, Optional
from uuid import uuid4

from a2a.types import AgentCard, Message, Task, TextPart
from pydantic import BaseModel, Field

from models.agent import Agent
from models.memory import ChatContext, RoomMemory
from models.room import Room, RoomAgentMessage, RoomMessage, RoomUserMessage
from models.task import BaseTask, MetaTask, TaskSession


class PaginationParams(BaseModel):
    page: int | None = Field(default=1, ge=1, description="Page number (1-indexed)")
    limit: int | None = Field(default=10, ge=1, le=100, description="Number of items per page")
    
    @property
    def skip(self) -> int:
        if not self.page:
            return 0
        return (self.page - 1) * self.limit


class FilterParams(BaseModel):
    # all nullable
    filters: dict[str, Any] | None = Field(default_factory=dict, description="MongoDB filter conditions")
    sort_by: str | None = Field(default=None, description="Field to sort by")
    sort_order: int | None = Field(default=-1, description="Sort order: 1 for ascending, -1 for descending")



class TaskRequest(BaseModel):
    task_id: str = Field(default_factory=lambda: str(uuid4()))
    query: str
    context: dict[str, Any] | None = Field(default_factory=dict)
    message: Message | None = None

    def to_message(self) -> Message:
        """Convert request to A2A protocol Message"""
        if self.message:
            return self.message

        # Create message if not provided
        parts = [TextPart(text=self.query)]
        return Message(role="user", parts=parts, metadata=self.context)


class AgentTaskRequest(BaseModel):
    task_id: str
    agent_id: str
    step_id: str
    input_data: Any
    context: dict[str, Any] | None = Field(default_factory=dict)
    message: Message | None = None

    def to_message(self) -> Message:
        """Convert agent task request to A2A protocol Message"""
        if self.message:
            return self.message

        # Create message from input data
        if isinstance(self.input_data, str):
            parts = [TextPart(text=self.input_data)]
        elif isinstance(self.input_data, dict) and "text" in self.input_data:
            parts = [TextPart(text=self.input_data["text"])]
        else:
            # Try to convert to string or use as-is
            try:
                text = str(self.input_data)
                parts = [TextPart(text=text)]
            except:
                # Use generic text if conversion fails
                parts = [TextPart(text=f"Processing step {self.step_id}")]

        # Add metadata
        metadata = {
            "task_id": self.task_id,
            "agent_id": self.agent_id,
            "step_id": self.step_id,
            **self.context,
        }

        return Message(role="user", parts=parts, metadata=metadata)


# for user
class UserInput(BaseModel):
    user_name: str
    user_input: str
    session_id: str | None = None


class InspectionCenterRequest(BaseModel):
    agent_id: str | None = None
    agent_url: str


class OrchestrationCenterRequest(BaseModel):
    task_id: str | None = None
    room_id: str | None = None
    room_user_message_id: str | None = None
    room_agent_message_id: str | None = None
    room_related_message_id: str | None = None


class DebatationCenterRequest(BaseModel):
    task_id: str


class AgentCenterRequest(BaseModel):
    agent_id: str | None = None
    agent_url: str | None = None
    agent_card: AgentCard | None = None
    call_increment: int | None = 0
    call_success_increment: int | None = 0
    like_increment: int | None = 0
    dislike_increment: int | None = 0
    query_text: str | None = None
    agent: Agent | None = None
    agent_count: int | None = 0


class BaseAgent(BaseModel):
    agent_url: str | None = None
    agent_card: AgentCard | None = None
    call_increment: int | None = 0
    call_success_increment: int | None = 0
    like_increment: int | None = 0
    dislike_increment: int | None = 0
    query_text: str | None = None
    agent: Agent | None = None
    agent_count: int | None = 0


class AgentCreate(BaseAgent):
    agent_id: str | None = Field(
        default_factory=lambda: str(uuid4()),
        pattern=r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$',
        description="Must be a valid UUID string"
    )
    agent_url: str
    agent_card: AgentCard


class AgentUpdate(BaseAgent):
    agent_id: str | None
    # agent_url: str | None
    agent_card: AgentCard | None
    call_count: int | None
    call_success_count: int | None
    like_count: int | None
    dislike_count: int | None
    agent_status: str | None


class AgentPatch(BaseAgent):
    agent_id: str


class TaskCenterRequest(BaseModel):
    task_id: str | None = None
    user_name: str | None = None
    parent_task_id: str | None = None
    session_id: str | None = None
    agent_id: str | None = None
    meta_task: MetaTask | None = None
    base_task: BaseTask | None = None
    task_session: TaskSession | None = None
    task: Task | None = None
    message: Message | None = None
    user_input: str | None = None
    execution_order: int = 0
    depends_on_tasks: list[str] | None = None
    context_from_previous: dict[str, Any] | None = None


class ChatRequest(BaseModel):
    user_name: str
    user_input: str
    session_id: str | None = None

class ChatMemoryRequest(BaseModel):
    user_name: str | None = None
    session_id: str | None = None
    user_input: str | None = None
    agent_response: str | None = None
    chat_context: ChatContext | None = None


class RoomCenterRoomSettingRequest(BaseModel):
    room_id: str | None = None
    room_name: str | None = None
    room_owner_id: str | None = None
    room_owner_name: str | None = None
    room_agent_set: Dict[str, str] | None = None
    room_created_at: datetime | None = None
    extend_info: dict[str, Any] | None = None
    room: Room | None = None


class RoomCenterUserMessageRequest(BaseModel):
    room_id: str | None = None
    message_id: str | None = None
    related_message_id: Optional[str] = None
    user_id: str | None = None
    user_name: str | None = None
    user_input: str | None = None
    message_created_at: datetime | None = None
    extend_info: dict[str, Any] | None = None
    message: RoomUserMessage | None = None

class RoomCenterAgentMessageRequest(BaseModel):
    room_id: str | None = None
    message_id: str | None = None
    related_message_id: Optional[str] = None
    agent_id: str | None = None
    agent_name: str | None = None
    agent_message_content: Task | None = None
    message_created_at: datetime | None = None
    extend_info: dict[str, Any] | None = None
    message: RoomAgentMessage | None = None

class RoomCenterMemoryRequest(BaseModel):
    room_id: str | None = None
    memory_id: str | None = None
    memory_content: str | None = None
    memory_created_at: datetime | None = None
    extend_info: dict[str, Any] | None = None
    memory: RoomMemory | None = None

class RoomCenterRoomMessageRequest(BaseModel):
    room_id: str | None = None
    message_id: str | None = None
    message_type: str | None = None
    message_content: str | None = None
    message_created_at: datetime | None = None
    extend_info: dict[str, Any] | None = None
    message: RoomMessage | None = None

File: frontend/src/lib/types/request.ts
/**
 * Request types for the backend API.
 *
 * A2A protocol types are imported from the canonical @a2a-js/sdk package.
 * Only backend-specific request wrappers and orchestration types are defined here.
 */

// ── Re-export A2A types needed by downstream consumers of this module ────
export type {
  TaskState,
  Part,
  TextPart,
  FilePart,
  DataPart,
  FileWithBytes,
  FileWithUri,
  Message,
  Task,
  TaskStatus,
  Artifact,
  AgentCard,
  AgentCapabilities,
  AgentExtension,
  AgentInterface,
  AgentProvider,
  AgentSkill,
  AgentCardSignature,
  SecurityScheme,
  APIKeySecurityScheme,
  HTTPAuthSecurityScheme,
  OAuth2SecurityScheme,
  OpenIdConnectSecurityScheme,
  MutualTLSSecurityScheme,
  OAuthFlows,
  AuthorizationCodeOAuthFlow,
  ClientCredentialsOAuthFlow,
  ImplicitOAuthFlow,
  PasswordOAuthFlow,
} from '@a2a-js/sdk'

import type {
  AgentCard,
  Task,
  Message,
} from '@a2a-js/sdk'

import type { Room } from './response'

export type { Room }

/**
 * The location of the API key.
 * (Not exported from @a2a-js/sdk as a standalone type.)
 */
export type In = "cookie" | "header" | "query";

/**
 * Identifies the sender of the message.
 * (Not exported from @a2a-js/sdk as a standalone type.)
 */
export type Role = "agent" | "user";

export type AgentStatus = "active" | "inactive" | "deleted";

export interface Agent {
  agent_id: string;
  provider_id?: string | null;
  agent_card: AgentCard;
  public_url?: string | null;
  agent_status?: AgentStatus | null;
  call_count?: number;
  call_success_count?: number;
  like_count?: number;
  dislike_count?: number;
}
export interface AgentCenterRequest {
  agent_id?: string | null;
  agent_url?: string | null;
  provider_id?: string | null;
  agent_card?: AgentCard | null;
  call_increment?: number | null;
  call_success_increment?: number | null;
  like_increment?: number | null;
  dislike_increment?: number | null;
  query_text?: string | null;
  agent?: Agent | null;
  agent_count?: number | null;
}
export interface APIKeyCreateRequest {
  name: string;
}
export interface AgentCreate {
  agent_url: string;
  agent_card: AgentCard;
  call_count?: number | null;
  call_success_count?: number | null;
  like_count?: number | null;
  dislike_count?: number | null;
  agent_status?: AgentStatus | null;
  /**
   * Must be a valid UUID string
   */
  agent_id?: string | null;
}
export interface AgentGroupCreateRequest {
  name: string;
  description?: string | null;
  owner_id: string;
  agents?: string[];
}
export interface AgentGroupRequest {
  group_id?: string | null;
  name?: string | null;
  description?: string | null;
  owner_id?: string | null;
  agents?: string[] | null;
}
export interface AgentGroupUpdateRequest {
  group_id: string;
  name?: string | null;
  description?: string | null;
  agents?: string[] | null;
}
export interface AgentPatch {
  agent_url?: string | null;
  agent_card?: AgentCard | null;
  call_count?: number | null;
  call_success_count?: number | null;
  like_count?: number | null;
  dislike_count?: number | null;
  agent_status?: AgentStatus | null;
}
export interface AgentTaskRequest {
  task_id: string;
  agent_id: string;
  step_id: string;
  input_data: unknown;
  context?: {
    [k: string]: unknown;
  } | null;
  message?: Message | null;
}
export interface AgentUpdate {
  agent_url?: string | null;
  agent_card: AgentCard | null;
  call_count: number | null;
  call_success_count: number | null;
  like_count: number | null;
  dislike_count: number | null;
  agent_status: AgentStatus | null;
  agent_id: string | null;
}
export interface BaseAgent {
  agent_url?: string | null;
  agent_card?: AgentCard | null;
  call_count?: number | null;
  call_success_count?: number | null;
  like_count?: number | null;
  dislike_count?: number | null;
  agent_status?: AgentStatus | null;
}
/**
 * A BaseTask represents a complete user request and serves as the top-level container.
 * It wraps a Task object and includes session/user metadata for tracking purposes.
 * This is the main task that gets decomposed into MetaTasks for multi-agent processing.
 */
export interface BaseTask {
  task_id: string;
  session_id: string;
  user_name: string;
  task: Task;
  extend_info?: unknown;
}
export interface ChatRequest {
  user_name: string;
  user_input: string;
  session_id?: string | null;
}
export interface DebatationCenterRequest {
  task_id: string;
}
export interface FilterParams {
  /**
   * MongoDB filter conditions
   */
  filters?: {
    [k: string]: unknown;
  } | null;
  /**
   * Field to sort by
   */
  sort_by?: string | null;
  /**
   * Sort order: 1 for ascending, -1 for descending
   */
  sort_order?: number | null;
}
export interface InspectionCenterRequest {
  agent_id?: string | null;
  agent_url: string;
}
/**
 * A MetaTask represents an atomic subtask created from decomposing a larger user request(BaseTask).
 * These are the individual work units assigned to specific agents in the multi-agent system.
 * Each MetaTask contains a Task object with the actual agent communication data.
 */
export interface MetaTask {
  task_id: string;
  parent_task_id: string;
  agent_id?: string;
  task_description?: string | null;
  task?: Task | null;
  execution_order?: number;
  depends_on_tasks?: string[] | null;
  context_from_previous?: {
    [k: string]: unknown;
  } | null;
  extend_info?: unknown;
}
export interface OrchestrationRequest {
  task_id?: string | null;
  room_id?: string | null;
  room_user_message_id?: string | null;
  room_agent_message_id?: string | null;
  room_related_message_id?: string | null;
}
export interface PaginationParams {
  /**
   * Page number (1-indexed)
   */
  page?: number | null;
  /**
   * Number of items per page
   */
  limit?: number | null;
}
export interface RoomAgentMessage {
  room_id: string;
  message_id: string;
  message_created_at?: string;
  message_type?: string;
  user_id?: string | null;
  agent_id?: string | null;
  related_message_id?: string | null;
  message_content: MessageContent;
  extend_info?: unknown;
}
export interface MessageContent {
  message_text?: string | null;
  message_task?: Task | null;
  [k: string]: unknown;
}
export interface RoomCenterAgentMessageRequest {
  room_id?: string | null;
  message_id?: string | null;
  related_message_id?: string | null;
  agent_id?: string | null;
  agent_name?: string | null;
  agent_message_content?: Task | null;
  message_created_at?: string | null;
  extend_info?: {
    [k: string]: unknown;
  } | null;
  message?: RoomAgentMessage | null;
}
export interface RoomCenterMemoryRequest {
  room_id?: string | null;
  memory_id?: string | null;
  memory_content?: string | null;
  memory_created_at?: string | null;
  extend_info?: {
    [k: string]: unknown;
  } | null;
  memory?: RoomMemory | null;
  room_agent_set?: {
    [k: string]: string;
  } | null;
  user_id?: string | null;
}
export interface RoomMemory {
  room_id: string;
  memory_id: string;
  memory_content?: MemoryContent;
  memory_created_at?: string;
  extend_info?: unknown;
}
/**
 * Room conversation memory with structured history.
 * Similar to ChatGPT/Claude conversation context management.
 */
export interface MemoryContent {
  summary?: string | null;
  conversation_history?: ConversationTurn[];
  memory_text?: string | null;
  [k: string]: unknown;
}
/**
 * A single turn in the conversation (ChatGPT/Claude style).
 * Represents either a user message or an agent response.
 */
export interface ConversationTurn {
  role: "user" | "agent";
  content: string;
  agent_id?: string | null;
  agent_name?: string | null;
  user_id?: string | null;
  timestamp?: string;
  [k: string]: unknown;
}
export interface RoomCenterRoomMessageRequest {
  room_id?: string | null;
  limit?: number | null;
  cursor?: string | null;
  message_id?: string | null;
  message_type?: string | null;
  message_content?: string | null;
  message_created_at?: string | null;
  extend_info?: {
    [k: string]: unknown;
  } | null;
  message?: RoomMessage | null;
}
/**
 * Unified room message format for both user and agent messages
 */
export interface RoomMessage {
  room_id: string;
  message_id: string;
  message_created_at?: string;
  message_type: string;
  user_id?: string | null;
  agent_id?: string | null;
  related_message_id?: string | null;
  message_content: MessageContent;
}
export interface RoomCenterRoomSettingRequest {
  room_id?: string | null;
  room_name?: string | null;
  room_owner_id?: string | null;
  room_owner_name?: string | null;
  room_created_at?: string | null;
  extend_info?: {
    [k: string]: unknown;
  } | null;
  room?: Room | null;

  // ── Legacy fields (accepted during rollout, canonical wins) ────────────
  room_agent_set?: {
    [k: string]: string;
  } | null;
  /** @deprecated Use canonical membership_seed_input instead. */
  applied_from_group?: string | null;

  // ── Canonical membership write input (mutually exclusive union) ────────
  membership_seed_input?: "manual" | "saved_group" | "all_current_agents";
  room_agent_ids?: string[];
  seed_group_id?: string;
  seed_all_current_agents?: true;
}
export interface RoomCenterUserMessageRequest {
  room_id?: string | null;
  message_id?: string | null;
  related_message_id?: string | null;
  user_id?: string | null;
  user_name?: string | null;
  user_input?: string | null;
  message_created_at?: string | null;
  extend_info?: {
    [k: string]: unknown;
  } | null;
  message?: RoomUserMessage | null;
  client_request_id?: string | null;
}

type SendMessageBasePayload = {
  message: unknown
  client_request_id: string
}

export type SendMessagePayload = SendMessageBasePayload & (
  | {
      mentioned_agent_ids: [string, ...string[]]
      message_target_mode?: never
      target_group_id?: never
    }
  | {
      message_target_mode: 'room_default' | 'all_agents'
      mentioned_agent_ids?: never
      target_group_id?: never
    }
  | {
      message_target_mode: 'saved_group'
      target_group_id: string
      mentioned_agent_ids?: never
    }
)

export interface RoomUserMessage {
  room_id: string;
  message_id: string;
  message_created_at?: string;
  message_type?: string;
  user_id?: string | null;
  agent_id?: string | null;
  related_message_id?: string | null;
  message_content: MessageContent;
  extend_info?: unknown;
}
export interface TaskCenterRequest {
  task_id?: string | null;
  user_name?: string | null;
  parent_task_id?: string | null;
  session_id?: string | null;
  agent_id?: string | null;
  meta_task?: MetaTask | null;
  base_task?: BaseTask | null;
  task_session?: TaskSession | null;
  task?: Task | null;
  message?: Message | null;
  user_input?: string | null;
  execution_order?: number;
  depends_on_tasks?: string[] | null;
  context_from_previous?: {
    [k: string]: unknown;
  } | null;
}
/**
 * A TaskSession represents a chat conversation between a user and the multi-agent system.
 * It tracks session metadata like creation time, user info, and session description.
 * Multiple BaseTask objects can belong to one TaskSession during a conversation.
 */
export interface TaskSession {
  session_id: string;
  user_name: string;
  session_name: string;
  session_description?: string | null;
  session_created_at?: string;
  session_updated_at?: string;
  extend_info?: unknown;
}
export interface TaskRequest {
  task_id?: string;
  query: string;
  context?: {
    [k: string]: unknown;
  } | null;
  message?: Message | null;
}
export interface UserInput {
  user_name: string;
  user_input: string;
  session_id?: string | null;
}
